### PR TITLE
feat(input_pipeline): Add support for chunking long sequences instead truncation

### PR DESCRIPTION
**Pull request by bzantium**
This PR introduces "chunking" as an alternative to "truncation" in the Grain input pipeline.

Previously, the `TokenizeAndTrim` operation (`MapTransform`) would truncate any document longer than `max_target_length`, discarding all subsequent tokens. This change introduces a new `TokenizeAndChunk` operation (`FlatMapTransform`) that splits a single long document into *multiple* training examples, each no longer than `max_target_length`.

This new behavior is controlled by a new configuration flag, `use_truncation`.

  - **Why is this change being made?**
    The default truncation behavior is highly data-inefficient for corpora with many long documents (like C4). It wastes significant amounts of data, compute, and storage, and may bias the model by only ever training on the beginning of documents.

  - **The problem being solved and any relevant context:**
    This PR solves the problem of data loss during tokenization for long sequences. By using a 1:N `FlatMapTransform`, we can map one long input document to a list of multiple, valid training chunks, ensuring 100% of the tokenized data is used.

  - **Why this is a good solution:**
    This solution is efficient and flexible. It utilizes the `FlatMapTransform` provided by Grain, which is designed for this 1:N mapping. It is also fully backwards-compatible, as the new chunking behavior is "opt-in" by setting `use_truncation = False` in the config. The default behavior remains truncation.

  - **Some information about the specific implementation:**

    1.  **`_grain_tokenizer.py`**: A new `TokenizeAndChunk` class has been added. It inherits from `grain.experimental.FlatMapTransform` and implements the `flat_map` method to split a list of token IDs into multiple chunks.
    2.  **`_grain_data_processing.py`**: The `pretrain_preprocessing_pipeline` function has been updated with a conditional check for `config.use_truncation`:
          - If `True`, it uses the existing `dataset.map(TokenizeAndTrim(...))`.
          - If `False`, it uses `dataset.apply(TokenizeAndChunk(...))`.
    3.  **Requirement:** The `dataset.apply()` method and support for `FlatMapTransform` are recent features in Grain. This PR requires a version of Grain installed directly from the main branch.
        ```bash
        pip install git+https://github.com/google/grain.git
        ```

  - **Shortcomings of the solution and possible future improvements.**
    The `max_fan_out` attribute in `TokenizeAndChunk` is set with a class-level default (`2048`). If a document is exceptionally long and produces more chunks than this, it will error. This could be exposed as a configuration option in the future if needed.

### Tests

This change is tested with a new, self-contained unit test file: `tests/tokenizer_transform_test.py`.

  - This test **does not** require real data (like C4) or JAX/TPU.
  - It uses a `MockTokenizer` to provide known, deterministic tokenization (`"a b c" -> [1, 2, 3]`).
  - It uses an in-memory `grain.MapDataset.source` with a small, known dataset to test edge cases (short text, long text, and multi-chunk text).
  - Four separate test cases were added to verify the logic:
    1.  **`test_tokenize_and_trim`**: Verifies the original 1:1 truncation logic is correct.
    2.  **`test_tokenize_and_chunk`**: Verifies the new 1:N chunking logic (e.g., an input with 7 tokens and `max_len=5` correctly produces two new examples with 5 and 2 tokens).
    3.  **`test_trim_and_pad_chaining`**: Verifies that the output of `TokenizeAndTrim` can be correctly chained into a subsequent `PadToMaxLength` transform.
    4.  **`test_chunk_and_pad_chaining`**: Verifies that *all* outputs from `TokenizeAndChunk` are correctly chained into `PadToMaxLength` (e.g., both the 5-token chunk and the 2-token chunk are correctly padded).

To reproduce, you can run the new test file directly:

```bash
python -m unittest tests/tokenizer_transform_test.py
```

Fixes: #2344 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/MaxText/configs/base.yml [hunk 1]
```diff
@@ -458,6 +458,7 @@ tokenize_train_data: True  # False if the dataset is pre-tokenized
 tokenize_eval_data: True  # False if the dataset is pre-tokenized
 add_bos: True
 add_eos: True
+use_truncation: True # If False, use chunking for long sequences instead of truncation
```
Comment (aireenmei) [inline src/MaxText/configs/base.yml]: Could you point to the class for users who are interested in the detailed implementation.

File: src/MaxText/input_pipeline/_grain_tokenizer.py [hunk 2]
```diff
@@ -64,3 +64,64 @@ def __setstate__(self, state):
...
+      return []
+
+    output_elements = []
+
+    for i in range(0, len(token_ids), max_len):
+      chunk = token_ids[i : i + max_len]
+      chunk_np = np.asarray(chunk, dtype=np.int32)
+      new_element = element.copy()
+
+      for feature_name in self.feature_names:
```
Comment (aireenmei) [inline src/MaxText/input_pipeline/_grain_tokenizer.py]: Replacing all the features with the chunked primary_feature can be a bit confusing here, and complicates this module. I think this way is cleaner: TokenizeAndChunk takes in a primary_feature param and only change that column, leaving all other features unchanged (copied to all chunks). Then in _grian_data_processing.py, move the rekey block (https://github.com/AI-Hypercomputer/maxtext/blob/main/src/MaxText/input_pipeline/_grain_data_processing.py#L98-L100) to after tokenization to map 'text' to 'inputs' and 'targets'. What do you think?

**Review by aireenmei — COMMENTED** [review]
Thanks for the feature! And great unit tests! Just some minor comments.

**Comment by bzantium** [comment]
Thanks for the great feedback!

I've pushed the changes addressing all your points:

* **`base.yml` comment**: Added the comment to the `use_truncation` flag to point to the implementation classes as you suggested.
* **Tokenizer Refactoring**:
    * I've simplified both `TokenizeAndTrim` and `TokenizeAndChunk` to operate on a single `text_column`.
    * The `Rekey` transform has been moved to execute *after* the tokenization step for both code paths, which cleans up the pipeline logic nicely.

to: @aireenmei

File: src/MaxText/input_pipeline/_grain_data_processing.py [hunk 3]
```diff
@@ -95,9 +95,9 @@ def pretrain_preprocessing_pipeline(dataset, config, data_columns, tokenize, gra
     dataset = dataset.map(_input_pipeline_utils.NormalizeFeatures(data_columns, tokenize))
 
   assert len(data_columns) == 1
-  rekey_dict = {"inputs": "text", "targets": "text"}
-  dataset = dataset.map(_input_pipeline_utils.Rekey(rekey_dict))
+  text_column = "text"
```
Comment (aireenmei) [inline src/MaxText/input_pipeline/_grain_data_processing.py]: we can have text_column = data_columns[0]? Let's move the rekey_dict definition to right before the Rekey op in line 131.

File: src/MaxText/input_pipeline/_grain_tokenizer.py [hunk 4]
```diff
@@ -28,30 +28,26 @@ class TokenizeAndTrim(grain.MapTransform):
   """Tokenize and trim features to sequence length."""
 
   # pylint: disable=attribute-defined-outside-init
-  feature_names: str | Sequence[str]
```
Comment (aireenmei) [inline src/MaxText/input_pipeline/_grain_tokenizer.py:31]: Let's keep TokenizeAndTrim as it was to allow supporting multiple columns, I think it's needed for the DPO support (https://github.com/AI-Hypercomputer/maxtext/pull/947)

File: src/MaxText/input_pipeline/_grain_data_processing.py [hunk 5]
```diff
@@ -115,11 +115,21 @@ def pretrain_preprocessing_pipeline(dataset, config, data_columns, tokenize, gra
...
-        )
-    )
+    if config.use_truncation:
+      dataset = dataset.map(
+          _grain_tokenizer.TokenizeAndTrim(
+              text_column, config.max_target_length, config.add_bos, config.add_eos, tokenizer_model
+          )
+      )
+    else:
+      dataset = dataset.apply(
```
Comment (SurbhiJainUSC) [inline src/MaxText/input_pipeline/_grain_data_processing.py]: What is the reason for using `dataset.apply` and not `dataset.map` similar to how `TokenizeAndTrim` uses at line 119?

**Comment by aireenmei** [comment]
Looks like the github actions tests need to be triggered by a maintainer. Please take a look at the test failures. You can also run them locally

File: src/MaxText/input_pipeline/_grain_tokenizer.py [hunk 6]
```diff
@@ -64,3 +60,55 @@ def __setstate__(self, state):
     self.__dict__.update(state)
     self._processor = None
     self._initialize_processor_lock = threading.Lock()
+
+
+@dataclasses.dataclass
```
Comment (SurbhiJainUSC) [inline src/MaxText/input_pipeline/_grain_tokenizer.py:68]: Most of the code looks similar to `TokenizeAndTrim` class. Can we create a base class, something like `TokenizerTransform` and then `TokenizeAndTrim` and `TokenizeAndChunk` can inherit from that class?

File: src/MaxText/input_pipeline/_grain_tokenizer.py (hunk 4 above)
Comment (bzantium) [reply src/MaxText/input_pipeline/_grain_tokenizer.py:31]: You're right. I will roll back TokenizeAndTrim and TokenizeAndChunk can consume multiple columns.

File: src/MaxText/input_pipeline/_grain_data_processing.py (hunk 5 above)
Comment (bzantium) [reply src/MaxText/input_pipeline/_grain_data_processing.py]: That is because of `MapDataset`'s `map` only takes 1:1 mapping transformations (return `MapMapDataset`) but `apply` supports 1:N transformations (return `FlatMapMapDataset`). You can find it here: https://github.com/google/grain/blob/main/grain/_src/python/dataset/transformations/map.py
I also want it to support this with `map` for consistency 😂.
```py
  def map(
      self, transform: transforms.MapTransform | Callable[[T], S]
  ) -> MapDataset[S]:
    """Returns a dataset containing the elements transformed by ``transform``.

    Example usage::

      ds = MapDataset.range(5)
      ds = ds.map(lambda x: x + 10)
      list(ds) == [10, 11, 12, 13, 14]

    Args:
      transform: Either a ``MapTransform`` containing the ``map`` method or a
        callable that takes an element and returns a new element.

    Returns:
      A dataset containing the elements of the original dataset transformed by
      ``transform``.
    """
    # Loaded lazily due to a circular dependency (dataset <-> map).
    # pylint: disable=g-import-not-at-top
    from grain._src.python.dataset.transformations import (
        map as map_dataset,
    )
    # pylint: enable=g-import-not-at-top
    return map_dataset.MapMapDataset(parent=self, transform=transform)
```
```py
  def apply(
      self,
      transformations: transforms.Transformation | transforms.Transformations,
  ) -> IterDataset:
    """Returns a dataset with the given transformation(s) applied.

    Syntactic sugar to avoid dispatch by transformation type.

    Example usage::

      ds = grain.MapDataset.range(5).to_iter_dataset()
      ds = ds.apply([AddOne(), grain.transforms.Batch(2)])
      list(ds) == [np.ndarray([1, 2]), np.ndarray([3, 4]), np.ndarray([5])]

    Args:
      transformations: one or more transformations to apply.

    Returns:
      Dataset with the given transformations applied.
    """
    return apply_transformations(self, transformations)

def apply_transformations(
    ds: _ConsistentDatasetType,
    transformations: transforms.Transformation | transforms.Transformations,
) -> _ConsistentDatasetType:
  """Applies transformations to a dataset.

  DEPRECATED: Use `ds.apply(transformations)` instead.

  Args:
    ds: `MapDataset` or `IterDataset` to apply the transformations to.
    transformations: one or more transformations to apply.

  Returns:
    Dataset of the same type with transformations applied.
  """
  if not isinstance(transformations, Sequence):
    transformations = (transformations,)
  for transformation in transformations:
    match transformation:
      case transforms.Batch():
        ds = ds.batch(
            transformation.batch_size,
            drop_remainder=transformation.drop_remainder,
        )
      case transforms.MapTransform():
        ds = ds.map(transformation)
      case transforms.RandomMapTransform():
        ds = ds.random_map(transformation)
      case transforms.MapWithIndex():
        ds = ds.map_with_index(transformation)
      case transforms.FlatMapTransform():
        # Loaded lazily due to a circular dependency (dataset <-> flatmap).
        # pylint: disable=g-import-not-at-top
        from grain._src.python.dataset.transformations import flatmap
        # pylint: enable=g-import-not-at-top
        if isinstance(ds, MapDataset):
          ds = flatmap.FlatMapMapDataset(ds, transformation)
        else:
          ds = flatmap.FlatMapIterDataset(ds, transformation)
      case transforms.Filter():
        ds = ds.filter(transformation)
      case _:
        raise NotImplementedError(
            f"Transformation type: {transformation} is not supported."
        )
  return ds
  ```

File: src/MaxText/input_pipeline/_grain_tokenizer.py (hunk 6 above)
Comment (bzantium) [reply src/MaxText/input_pipeline/_grain_tokenizer.py:68]: Yeah, that would be better!

File: src/MaxText/input_pipeline/_grain_data_processing.py (hunk 3 above)
Comment (bzantium) [reply src/MaxText/input_pipeline/_grain_data_processing.py]: Got it. I've updated the variable to `text_column` and moved the `rekey_dict` as requested. The changes have been pushed.

File: src/MaxText/input_pipeline/_grain_tokenizer.py (hunk 4 above)
Comment (bzantium) [reply src/MaxText/input_pipeline/_grain_tokenizer.py:31]: Thanks for the context about the DPO support. I've reverted `TokenizeAndTrim` back to its original implementation (but refactored as recommended below) to support multiple columns, and I've pushed the change.

**Review by aireenmei — APPROVED** [review]
LGTM, thanks!

**Comment by bzantium** [comment]
@aireenmei yeah, I think all look good for now. thanks for the detailed reviews!

**Comment by bzantium** [comment]
I think `dataset.apply` requires latest `grain` so need to change requirements.txt like grain[parquet] @ https://github.com/google/grain/archive/bab58eabf0c94d16002b5d82dda8d8320edd3c7b.zip)

**Comment by aireenmei** [comment]
> I think `dataset.apply` requires latest `grain` so need to change requirements.txt like grain[parquet] @ https://github.com/google/grain/archive/bab58eabf0c94d16002b5d82dda8d8320edd3c7b.zip)

The github runner uses requirements_with_jax_ai_image.txt. Could you change it with the PR and see if that fix the test?

**Comment by Rohan-Bierneni** [comment]
Hi @bzantium can you rebase your pr and run the tests again. There were some changes to testing infra last week that require a rebase for successful image builds in unit tests.

**Comment by bzantium** [comment]
@aireenmei @Rohan-Bierneni @SurbhiJainUSC 
I revised the code to use `WithOptionsIterDataset` and `apply_transformations` instead of `dataset.apply` because `grain` needs to be pre-compiled because it generates python file using pb2 with compilation but installation with latest commit does not perform this (pypi supports this so I decided to use grain==0.2.12).
Moreover, I found that add_bos/eos are used when building tokenizer not on transformation so I removed those arguments. I've checked current code works fine on my TPU cluster.

**Comment by aireenmei** [comment]
> @aireenmei @Rohan-Bierneni @SurbhiJainUSC I revised the code to use `WithOptionsIterDataset` and `apply_transformations` instead of `dataset.apply` because `grain` needs to be pre-compiled because it generates python file using pb2 with compilation but installation with latest commit does not perform this (pypi supports this so I decided to use grain==0.2.12). Moreover, I found that add_bos/eos are used when building tokenizer not on transformation so I removed those arguments. I've checked current code works fine on my TPU cluster.

Thanks for the update! If grain==0.2.12 doesn't work, let me know if we need to request a new release from the grain team.

**Comment by bzantium** [comment]
@aireenmei thanks for the fast reply! I've checked current implementation works fine with grain==0.12.2 on tpu-v6e but I think if it's possible for grain team to release new version, that would be nice in terms of this PR (can make code more neat as previous) and other upcoming features that I want to add soon.

**Comment by bzantium** [comment]
https://pypi.org/project/grain/0.2.13/
new grain version is now available! I will test with the new code and push soon.

**Comment by bzantium** [comment]
I've successfully tested this code on my tpus... why test failure happened...?!

**Comment by aireenmei** [comment]
The corresponding tpu test are passing for MaxText head. Could you rebase again? Maxtext gpu unit test seems to not pick up the new Grain version. But at head it has picked up the newest one (https://github.com/AI-Hypercomputer/maxtext/actions/runs/18940014406/job/54076248442). I manually added pull ready label, not sure if this can make the test run automatically once you rebase and repush the branch

**Comment by bzantium** [comment]
@aireenmei 
Thanks for checking. I also saw in the logs that the failed test was using `grain==0.2.12`.
I've just rebased the branch and modified the `requirements_with_jax_stable_stack_0_6_1_pipreqs.txt` file. 
It seems the tests didn't run automatically after the rebase.

**Comment by aireenmei** [comment]
Tests are looking good. Could you fix the linter error? The script code_style.sh maybe helpful. Then squash the commits into one.

**Comment by bzantium** [comment]
@aireenmei I think it might be fine this time.

**Review by parambole — APPROVED** [review]

**Comment by aireenmei** [comment]
@bzantium Could you squash the 22 commits into one? MaxText allows merging 1 commit per PR

**Comment by bzantium** [comment]
@aireenmei I squashed commits into one!
